### ISSUE TITLE
text spacing != 1.0 doesn't play nice with ligatures

### ISSUE THREAD

**Issue by jordanbrown0** [openscad/openscad#3859]
Text spacing is applied between Harfbuzz-generated glyphs.  In a font with ligatures, multiple letters can combine into one glyph, so the spacing does silly things.

```text("difficult", font="Palatino", spacing=2);```

yields

![image](https://user-images.githubusercontent.com/31949071/129524433-2433b051-aafb-466c-8e35-db83aa00f61b.png)

Perhaps when spacing is not 1.0 it could somehow disable use of ligatures.

**Comment by jordanbrown0** [comment]
Also, spacing is applied as a simple multiplication of the advance value, so you get more space between `M`s than between `I`s:

```text("MMMIII", spacing=2);```

![image](https://user-images.githubusercontent.com/31949071/129524701-8a618b3d-3682-4390-86f4-ca7dbdb997d3.png)

Setting the spacing to 2 does not mean "put twice as much space between glyphs"; instead, it means "advance twice as far as usual for each glyph".

Having spacing affect the inter-glyph spacing rather than the advance would probably yield more sensible results, but might be trickier since what comes out of Harfbuzz is the position of each glyph, not the space between them.  (And of course such a change would be an incompatible change.)

**Comment by jordanbrown0** [comment]
(I am **not** planning to address either of these behaviors in the text cleanup work that I'm doing with `textmetrics()`.)

**Comment by WillAdams** [comment]
I would argue that getting the ligature is providing what the user asked for --- if you don't want ligatures to be formed, turn them off in the text setting or insert a zero width no joiner (ZWNJ) in-between the characters to prevent the ligature.

**Comment by jordanbrown0** [comment]
Forming ligatures isn't bad.  It's the interaction between ligatures and the "spacing" parameter that is bad, or at least can easily produce unexpected results.

Between these two interactions, I think "spacing" ends up being almost useless, and should probably be replaced by a different feature with entirely different rules.

Workaround:  walk through the string character by character, measuring each character's size with textmetrics() and rendering it separately.

**Comment by t-paul** [comment]
FWIW, you don't want the ZWJ (`U+200D`), but the Zero Width **Non** Joiner (`U+200C`)
```
text("dif\u200Cf\u200Cicult", font="Palatino", spacing=2);
```
Still looks not very smooth though,
![grafik](https://user-images.githubusercontent.com/1330241/140947902-f2e62794-7342-483c-9b18-64fe4fe1c329.png)

But lets not forget monospaced fonts...
```
text("difficult", font="Liberation Mono:style=Regular", spacing=2);
```
![grafik](https://user-images.githubusercontent.com/1330241/140948322-4c5e9e8c-3c71-4daa-a2cb-4780ff389a86.png)

**Comment by WillAdams** [comment]
Have you tried using a monospace font such as Lucida Sans Typewriter?

**Comment by rizwansarwar** [comment]
> Have you tried using a monospace font such as Lucida Sans Typewriter?

Yes I have, of course it is bit better with Monospace. My use case is to make any font and calculate spacing using textmetrics.

**Comment by jordanbrown0** [comment]
@rizwansarwar You don't say what you mean by "consistent".  You cannot pull a string apart character by character and expect the same results that you would get from calling text() on the full string, because calling text() on the full string does kerning and ligature formation that can only be done when considering how the various characters relate to one another.

Here's a quick program that does a plausible job of evenly adding spacing between letters.  Use the customizer to experiment with its settings.
```
// String to display
s = "Hello, World!";
// Additional spacing (in OpenSCAD units)
spacing = 2;
// Font
font = "Liberation Sans";
// Font size
size = 10;

// Get the advance metrics (expected base location
// of the next glyph) for each character in the string,
// plus the desired additional spacing.
advances = [
    for (c = s)
        textmetrics(c, font=font, size=size).advance.x + spacing
];

// A function that adds successive entries in an array,
// so that runningSum([a,b,c]) yields [a, a+b, a+b+c].
function runningSum(a, i=0, sum=0) =
    i >= len(a)
    ? []
    : let(newSum = a[i]+sum)
        [ newSum, each runningSum(a, i+1, newSum) ];

// Prepend a zero, so that we have the positions of
// each character.
pos = concat(0, runningSum(advances));

// Position each character.
for (i=[0:len(s)-1])
    translate([pos[i],10,0]) text(s[i], font=font, size=size);
```

**Comment by jordanbrown0** [comment]
My original example (at the very top) is flawed:  I asked for "Palatino" but that's wrong; the font is "Palatino Linotype".  I ended up getting "Times New Roman", as calling `fontmetrics(font="Palatino")` shows (reformatted for readability):
```
ECHO: {
    nominal = { ascent = 12.3766; descent = -3.0043; };
    max = { ascent = 14.4382; descent = -4.2589; };
    interline = 15.9709;
    font = { family = "Times New Roman"; style = "Regular"; };
}
```
Nonetheless, it demonstrates the behavior I was describing.

**Comment by jordanbrown0** [comment]
Hmm.  As I experiment now, TNR does *not* demonstrate ligature formation.  I'm testing today on a different computer with a different Windows build; perhaps I *was* getting Palatino back in 2021.

**Comment by jordanbrown0** [comment]
I'm going to edit the comments to avoid leaving that false information.

**Comment by jordanbrown0** [comment]
For those trying to replicate the original experiment:  it appears that things have changed since then, probably in the transition from Windows 10 to Windows 11.  Today I do not have a font "Palatino"; the equivalent appears to be "Palatino Linotype".  Asking for "Palatino" yields "Times New Roman", which does not demonstrate the ligature.  Check with Help/Font List and/or fontmetrics(font="whatever") to be sure you're getting the font you expect.

[ Updated font in the original comment. ]

**Comment by jordanbrown0** [comment]
Setting spacing=0 and stacking the results can illustrate the spacing differences caused by kerning and ligatures.
```
translate([0,10,-1]) color("red") text(s, font=font, size=size);
```
In the examples below, the yellow is the "manually" spaced text resulting from the program above, with spacing=0.  The red is the normal `text()` spacing.

Ligatures in Palatino Linotype (combining `ffi` into one glyph):
![image](https://github.com/openscad/openscad/assets/31949071/6a008995-e7cf-43db-9f5c-bfc53055317c)
Kerning in Times New Roman (tucking the `o` into the space under the right side of the `W`):
![image](https://github.com/openscad/openscad/assets/31949071/ea74bd15-02b2-4749-8271-86459d7ac747)

**Comment by jordanbrown0** [comment]
Note that my "equal spacing" program above does not consider kerning, which is why the "o" in World is pushed so far to the right.

**Comment by jordanbrown0** [comment]
What it should perhaps do is accept a parameter that specifies extra spacing (in OpenSCAD units) to be added between characters, basically as is done in my example above.  This would necessarily sacrifice ligatures, but would also sacrifice kerning.  Perhaps there is a way to get Harfbuzz to disable ligatures; if so then the extra spacing could be applied between glyphs.

**Comment by jordanbrown0** [comment]
Probably this is mostly a documentation issue:  "don't expect reasonable results except for monospace fonts".

**Comment by amatulic** [comment]
How about instead of putting a disclaimer in the documentation (which is the easiest), the `spacing` parameter be fixed so that it still works with fixed-width fonts but inserts a reasonably-sized constant space between each character for proportional fonts?

Here's an example of something crude I just did, using the most common letter (lowercase "e") as the reference character to determine the difference in the `.advance` property when `spacing` is changed.

<img width="621" height="186" alt="Image" src="https://github.com/user-attachments/assets/3d6a6c25-d356-4d42-9733-217330e3d444" />

By coincidence, the total length of that word happened to be the same with the improved spacing. That wouldn't be the case with different strings, but the point is that you get spaced-out characters that are nicely distributed.

I get the same results if I use a zero as the reference character, which might be a more universal and consistent reference for any font.

My code to generate that graphic:
````
fontface = "Liberation Sans:style=Bold";
string = "Animals";
spacing = 2;
translate([0,15]) {
    color("black") text("spacing=1: ", size=5, font=fontface, halign="right");
    text(string, size=10, font=fontface);
}
color("black") text(str("spacing=",spacing," "), size=5, font=fontface, halign="right");
text(string, size=10, font=fontface, spacing=spacing);
translate([0,-15]) {
    color("black") text("Proposed: ", size=5, font=fontface, halign="right");
    text2(string, size=10, font=fontface, spacing=spacing);
}

module text2(text, size=10, font="Liberation Sans", direction="ltr", language="en", script="latin", halign="left", valign="baseline", spacing=1, em) {
    spc =  // space to add between each character
        textmetrics("e", 10, fontface, spacing=1).advance[0] * (spacing-1);
    ilast = len(string)-1;
    cpos = cumsum([
        for(i=[0:ilast]) textmetrics(text[i], size, font, direction, language, script, halign, valign, 1, em).advance[0]
    ], spc);
    echo(cpos);
    for(i=[0:ilast])
        translate([cpos[i], 0])
            text(text[i], size, font, direction, language, script, halign, valign, 1, em);
}

// for an array of numbers, return array of cumulative sums with a constant added to each element
function cumsum(a, adder, i=0, sum=[0]) = i>=len(a) ? sum :
cumsum(a, adder, i+1, concat(sum, sum[i]+a[i]+adder));
````

**Comment by jordanbrown0** [comment]
Seems plausible. Note that the algorithm that you suggest yields the current results for monospace fonts. (Probably you already knew that; the way I read what you wrote was that the two would need to be treated differently.)

**Comment by amatulic** [comment]
Yup, I knew that, and I intended to retain the current behavior with monospace fonts. Over a year ago I started working occasionally on better text handling in BOSL2 (including wordwrapping, autoscaling to a bounding box, font size units, etc.) and it got nearly done but I kept getting sidetracked by solving more immediate problems. Originally, I handled the font spacing issue similar to what I proposed above, but my attempt to retain compatibility with text() arguments was rejected in favor of making a whole new thing called write() that doesn't try to be compatible with text(), and handles spacing more like you did in your example. This morning I resurrected that write() code again with intent to raise a PR soon.

**Comment by jordanbrown0** [comment]
As noted above, one of the tricks is that to get nonstandard spacing you need to defeat ligature handling.  It looks like that's sort of possible to do in an option to Harfbuzz, though it appears that it's somewhat font-dependent.  Ref https://www.google.com/search?q=harfbuzz+disable+ligatures

Disabling ligature support, especially if you do it by busting the string apart into individual characters, would mean that spacing=1.00001 would yield significantly different results from spacing=1.0, and that would be kind of unfortunate.  (But I can't see any way around it.)

Note that in the general case busting a string into individual characters is fraught with peril, because "character" does not have an obvious definition in Unicode.  Unicode allows near-arbitrary composition of characters, so that (e.g.) u-with-umlaut can be represented as U+00FC LATIN SMALL LETTER U WITH DIAERESIS, or it can be represented as U+0075 LATIN SMALL LETTER U followed by U+0308 COMBINING DIAERESIS.  I believe that this and related issues are particularly bad for script languages like Arabic.

**Comment by amatulic** [comment]
>Note that in the general case busting a string into individual characters is fraught with peril, because "character" does not have an obvious definition in Unicode.

That's true. In the case of doing it in OpenSCAD, one would have to parse the string into individual "characters" that start with \\x, \\u, or \\U to run each one through textmetrics() to account for the individual character widths, but that's still complicated by the "combining" unicode characters. In my version of this for BOSL2 I also accounted for kerning between character pairs. I didn't do anything with ligatures or unicode characters though.
